src/controllers/wallet/tx-proposal/tx-proposal.controller.js

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const {
+  constants: hathorLibConstants,
+  helpersUtils,
+  SendTransaction,
+  transaction,
+  transaction: transactionUtils,
+  walletUtils,
+  storage,
+} = require('@hathor/wallet-lib');
+const { _ } = require('lodash');
+const { parametersValidation } = require('../../../helpers/validations.helper');
+const { lock, lockTypes } = require('../../../lock');
+const { cantSendTxErrorMessage } = require('../../../helpers/constants');
+const { mapTxReturn, getUtxosToFillTx } = require('../../../helpers/tx.helper');
+
+async function buildTxProposal(req, res) {
+  const validationResult = parametersValidation(req);
+  if (!validationResult.success) {
+    res.status(400).json(validationResult);
+    return;
+  }
+
+  // build a transaction hex
+  const { wallet } = req;
+  const network = wallet.getNetworkObject();
+  const { outputs } = req.body;
+
+  const tokens = new Map();
+  for (const output of outputs) {
+    if (!output.token) {
+      output.token = hathorLibConstants.HATHOR_TOKEN_CONFIG.uid;
+    }
+    if (!tokens.has(output.token)) {
+      tokens.set(output.token, { tokenUid: output.token, amount: 0 });
+    }
+    if (output.type === 'data') {
+      // The data output requires that the user burns 0.01 HTR
+      // this must be set here, in order to make the filter_address query
+      // work if the inputs are selected by this method
+      output.value = 1;
+    }
+    const sumObject = tokens.get(output.token);
+    sumObject.amount += output.value;
+  }
+
+  // Expects array of objects with {'txId', 'index'}
+  let inputs = req.body.inputs || [];
+  const changeAddress = req.body.change_address || null;
+
+  if (inputs.length > 0) {
+    if (inputs[0].type === 'query') {
+      const query = inputs[0];
+      inputs = [];
+
+      // query processing
+
+      for (const element of tokens) {
+        const [tokenUid, tokenObj] = element;
+        const queryOptions = {
+          ...query,
+          token: tokenUid,
+        };
+        const utxos = getUtxosToFillTx(wallet, tokenObj.amount, queryOptions);
+        if (!utxos) {
+          const response = {
+            success: false,
+            error: 'No utxos available for the query filter for this amount.',
+            token: tokenUid
+          };
+          res.send(response);
+          lock.unlock(lockTypes.SEND_TX);
```

Comment:
Why do you have the `unlock` if you don't have the `lock`?

Reply:
Removed.